<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>The Microbiome Hackers Guide</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="This is an <em>early draft</em> excerpt of a book for curious people with little or no knowledge of biology who want to learn the basics of the human microbiome in a hands-on, practical way — by experimenting on yourself. By practicing with your own examples, you’ll learn how to study and manipulate the microbes inside you.">
  <meta name="generator" content="bookdown 0.3 and GitBook 2.6.7">

  <meta property="og:title" content="The Microbiome Hackers Guide" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="cover.png" />
  <meta property="og:description" content="This is an <em>early draft</em> excerpt of a book for curious people with little or no knowledge of biology who want to learn the basics of the human microbiome in a hands-on, practical way — by experimenting on yourself. By practicing with your own examples, you’ll learn how to study and manipulate the microbes inside you." />
  <meta name="github-repo" content="richardsprague/z" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="The Microbiome Hackers Guide" />
  
  <meta name="twitter:description" content="This is an <em>early draft</em> excerpt of a book for curious people with little or no knowledge of biology who want to learn the basics of the human microbiome in a hands-on, practical way — by experimenting on yourself. By practicing with your own examples, you’ll learn how to study and manipulate the microbes inside you." />
  <meta name="twitter:image" content="cover.png" />

<meta name="author" content="Richard Sprague">


<meta name="date" content="2017-06-02">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="science.html">
<link rel="next" href="my-experiments.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Microbiome Hackers Guide</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Intro</a></li>
<li class="chapter" data-level="" data-path="preface.html"><a href="preface.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting Started</a><ul>
<li class="chapter" data-level="1.1" data-path="getting-started.html"><a href="getting-started.html#the-most-common-phyla"><i class="fa fa-check"></i><b>1.1</b> The most common phyla</a></li>
<li class="chapter" data-level="1.2" data-path="getting-started.html"><a href="getting-started.html#how-are-you-different"><i class="fa fa-check"></i><b>1.2</b> How are you different?</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="science.html"><a href="science.html"><i class="fa fa-check"></i><b>2</b> Learn the Science</a><ul>
<li class="chapter" data-level="2.1" data-path="science.html"><a href="science.html#microbes-everywhere"><i class="fa fa-check"></i><b>2.1</b> Microbes everywhere</a><ul>
<li class="chapter" data-level="2.1.1" data-path="science.html"><a href="science.html#microbes-in-the-air"><i class="fa fa-check"></i><b>2.1.1</b> Microbes in the air</a></li>
<li class="chapter" data-level="2.1.2" data-path="science.html"><a href="science.html#microbes-around-you"><i class="fa fa-check"></i><b>2.1.2</b> Microbes around you</a></li>
<li class="chapter" data-level="2.1.3" data-path="science.html"><a href="science.html#microbes-and-food"><i class="fa fa-check"></i><b>2.1.3</b> Microbes and food</a></li>
<li class="chapter" data-level="2.1.4" data-path="science.html"><a href="science.html#anti-microbials-in-food"><i class="fa fa-check"></i><b>2.1.4</b> Anti-microbials in food</a></li>
<li class="chapter" data-level="2.1.5" data-path="science.html"><a href="science.html#fermented-food"><i class="fa fa-check"></i><b>2.1.5</b> Fermented food</a></li>
<li class="chapter" data-level="2.1.6" data-path="science.html"><a href="science.html#fiber-and-your-gut-microbes"><i class="fa fa-check"></i><b>2.1.6</b> Fiber and your gut microbes</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="science.html"><a href="science.html#microbes-and-health"><i class="fa fa-check"></i><b>2.2</b> Microbes and health</a><ul>
<li class="chapter" data-level="2.2.1" data-path="science.html"><a href="science.html#how-much-is-toxic"><i class="fa fa-check"></i><b>2.2.1</b> How much is toxic?</a></li>
<li class="chapter" data-level="2.2.2" data-path="science.html"><a href="science.html#food-allergies"><i class="fa fa-check"></i><b>2.2.2</b> Food Allergies</a></li>
<li class="chapter" data-level="2.2.3" data-path="science.html"><a href="science.html#gluten"><i class="fa fa-check"></i><b>2.2.3</b> Gluten</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="science.html"><a href="science.html#an-intro-to-biology"><i class="fa fa-check"></i><b>2.3</b> An intro to biology</a><ul>
<li class="chapter" data-level="2.3.1" data-path="science.html"><a href="science.html#the-basics"><i class="fa fa-check"></i><b>2.3.1</b> The Basics</a></li>
<li class="chapter" data-level="2.3.2" data-path="science.html"><a href="science.html#taxonomy"><i class="fa fa-check"></i><b>2.3.2</b> Taxonomy</a></li>
<li class="chapter" data-level="2.3.3" data-path="science.html"><a href="science.html#microbiology"><i class="fa fa-check"></i><b>2.3.3</b> Microbiology</a></li>
<li class="chapter" data-level="2.3.4" data-path="science.html"><a href="science.html#human-anatomy"><i class="fa fa-check"></i><b>2.3.4</b> Human Anatomy</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html"><i class="fa fa-check"></i><b>3</b> Explore Your Microbiome</a><ul>
<li class="chapter" data-level="3.1" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#how-to-use-the-ubiome-taxonomy-files"><i class="fa fa-check"></i><b>3.1</b> How to Use the uBiome Taxonomy Files</a><ul>
<li class="chapter" data-level="3.1.1" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#study-your-results-in-excel"><i class="fa fa-check"></i><b>3.1.1</b> Study your results in Excel</a></li>
<li class="chapter" data-level="3.1.2" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#ubiome-json-taxonomy-files"><i class="fa fa-check"></i><b>3.1.2</b> uBiome JSON taxonomy files</a></li>
<li class="chapter" data-level="3.1.3" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#the-ubiome-data-tags"><i class="fa fa-check"></i><b>3.1.3</b> The uBiome data tags</a></li>
<li class="chapter" data-level="3.1.4" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#converting-ubiome-taxonomy-json-to-excel"><i class="fa fa-check"></i><b>3.1.4</b> Converting uBiome taxonomy JSON to Excel</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#comparing-samples"><i class="fa fa-check"></i><b>3.2</b> Comparing samples</a></li>
<li class="chapter" data-level="3.3" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#make-a-single-sheet-with-all-your-samples"><i class="fa fa-check"></i><b>3.3</b> Make a single sheet with all your samples</a></li>
<li class="chapter" data-level="3.4" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#analyze-your-ubiome-results-in-python"><i class="fa fa-check"></i><b>3.4</b> Analyze your uBiome Results in Python</a></li>
<li class="chapter" data-level="3.5" data-path="explore-your-microbiome.html"><a href="explore-your-microbiome.html#analyze-a-fastq-file"><i class="fa fa-check"></i><b>3.5</b> Analyze a FASTQ file</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="my-experiments.html"><a href="my-experiments.html"><i class="fa fa-check"></i><b>4</b> My Experiments</a><ul>
<li class="chapter" data-level="4.1" data-path="my-experiments.html"><a href="my-experiments.html#kefir"><i class="fa fa-check"></i><b>4.1</b> How kefir changes the microbiome</a></li>
<li class="chapter" data-level="4.2" data-path="my-experiments.html"><a href="my-experiments.html#traveling-in-mexico"><i class="fa fa-check"></i><b>4.2</b> Traveling in Mexico</a></li>
<li class="chapter" data-level="4.3" data-path="my-experiments.html"><a href="my-experiments.html#do-probiotics-work"><i class="fa fa-check"></i><b>4.3</b> Do Probiotics Work?</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="caseStudies.html"><a href="caseStudies.html"><i class="fa fa-check"></i><b>5</b> Case Studies</a><ul>
<li class="chapter" data-level="5.1" data-path="caseStudies.html"><a href="caseStudies.html#father-and-son"><i class="fa fa-check"></i><b>5.1</b> Father and son</a></li>
<li class="chapter" data-level="5.2" data-path="caseStudies.html"><a href="caseStudies.html#case-study-a-baby-microbiome"><i class="fa fa-check"></i><b>5.2</b> Case study: a baby microbiome</a></li>
<li class="chapter" data-level="5.3" data-path="caseStudies.html"><a href="caseStudies.html#alzheimers-and-the-microbiome"><i class="fa fa-check"></i><b>5.3</b> Alzheimers and the microbiome</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="next.html"><a href="next.html"><i class="fa fa-check"></i><b>6</b> Go Further</a><ul>
<li class="chapter" data-level="6.1" data-path="next.html"><a href="next.html#fermented-food-1"><i class="fa fa-check"></i><b>6.1</b> Fermented Food</a><ul>
<li class="chapter" data-level="6.1.1" data-path="next.html"><a href="next.html#yogurt"><i class="fa fa-check"></i><b>6.1.1</b> Yogurt</a></li>
<li class="chapter" data-level="6.1.2" data-path="next.html"><a href="next.html#pickling"><i class="fa fa-check"></i><b>6.1.2</b> Pickling</a></li>
<li class="chapter" data-level="6.1.3" data-path="next.html"><a href="next.html#whats-happening-in-sauerkraut"><i class="fa fa-check"></i><b>6.1.3</b> What’s happening in sauerkraut</a></li>
<li class="chapter" data-level="6.1.4" data-path="next.html"><a href="next.html#more-fermentation"><i class="fa fa-check"></i><b>6.1.4</b> More fermentation</a></li>
<li class="chapter" data-level="6.1.5" data-path="next.html"><a href="next.html#pickled-ginger"><i class="fa fa-check"></i><b>6.1.5</b> Pickled Ginger</a></li>
<li class="chapter" data-level="6.1.6" data-path="next.html"><a href="next.html#pickled-cucumbers"><i class="fa fa-check"></i><b>6.1.6</b> Pickled cucumbers</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="next.html"><a href="next.html#my-favorite-books-about-the-microbiome"><i class="fa fa-check"></i><b>6.2</b> My favorite books about the microbiome</a><ul>
<li class="chapter" data-level="6.2.1" data-path="next.html"><a href="next.html#knight-rob-follow-your-gut-how-the-ecosystem-in-your-gut-determines-your-health-mood-and-more"><i class="fa fa-check"></i><b>6.2.1</b> Knight, Rob <em>Follow Your Gut: How the Ecosystem in Your Gut Determines Your Health, Mood, and More</em></a></li>
<li class="chapter" data-level="6.2.2" data-path="next.html"><a href="next.html#sonnenburg-justin-the-good-gut-taking-control-of-your-weight-your-mood-and-your-long-term-health"><i class="fa fa-check"></i><b>6.2.2</b> Sonnenburg, Justin <em>The Good Gut: Taking Control of Your Weight, Your Mood, and Your Long-term Health</em></a></li>
<li class="chapter" data-level="6.2.3" data-path="next.html"><a href="next.html#other-books-worth-having-in-the-pile"><i class="fa fa-check"></i><b>6.2.3</b> Other Books Worth Having in the Pile</a></li>
<li class="chapter" data-level="6.2.4" data-path="next.html"><a href="next.html#not-recommended"><i class="fa fa-check"></i><b>6.2.4</b> Not Recommended</a></li>
<li class="chapter" data-level="6.2.5" data-path="next.html"><a href="next.html#other"><i class="fa fa-check"></i><b>6.2.5</b> Other</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="next.html"><a href="next.html#best-academic-papers"><i class="fa fa-check"></i><b>6.3</b> Best Academic Papers</a><ul>
<li class="chapter" data-level="6.3.1" data-path="next.html"><a href="next.html#popular-topics"><i class="fa fa-check"></i><b>6.3.1</b> Popular Topics</a></li>
<li class="chapter" data-level="6.3.2" data-path="next.html"><a href="next.html#academic-papers"><i class="fa fa-check"></i><b>6.3.2</b> Academic Papers</a></li>
<li class="chapter" data-level="6.3.3" data-path="next.html"><a href="next.html#general-overview"><i class="fa fa-check"></i><b>6.3.3</b> General Overview</a></li>
<li class="chapter" data-level="6.3.4" data-path="next.html"><a href="next.html#historic-papers"><i class="fa fa-check"></i><b>6.3.4</b> Historic Papers</a></li>
<li class="chapter" data-level="6.3.5" data-path="next.html"><a href="next.html#self-tracking"><i class="fa fa-check"></i><b>6.3.5</b> Self-tracking</a></li>
<li class="chapter" data-level="6.3.6" data-path="next.html"><a href="next.html#diet"><i class="fa fa-check"></i><b>6.3.6</b> Diet</a></li>
<li class="chapter" data-level="6.3.7" data-path="next.html"><a href="next.html#population-studies"><i class="fa fa-check"></i><b>6.3.7</b> Population studies</a></li>
<li class="chapter" data-level="6.3.8" data-path="next.html"><a href="next.html#enterotypes"><i class="fa fa-check"></i><b>6.3.8</b> Enterotypes</a></li>
<li class="chapter" data-level="6.3.9" data-path="next.html"><a href="next.html#large-population-summaries"><i class="fa fa-check"></i><b>6.3.9</b> Large population summaries</a></li>
<li class="chapter" data-level="6.3.10" data-path="next.html"><a href="next.html#methods"><i class="fa fa-check"></i><b>6.3.10</b> Methods</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="next.html"><a href="next.html#other-resources"><i class="fa fa-check"></i><b>6.4</b> Other Resources</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>7</b> Acknowledgments</a><ul>
<li class="chapter" data-level="7.1" data-path="acknowledgments.html"><a href="acknowledgments.html#style-guide"><i class="fa fa-check"></i><b>7.1</b> Style guide</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>8</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Microbiome Hackers Guide</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="explore-your-microbiome" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Explore Your Microbiome</h1>
<p>Your first step is to get some data. The most popular laboratory is a San Francisco startup called uBiome, that originally began as crowdfunded campaign on Indiegogo<a href="#fn17" class="footnoteRef" id="fnref17"><sup>17</sup></a>, but has since handled tens of thousands of samples from people like me who are curious to know what a 16S analysis will reveal about their own microbiome. For less than $100, you send them a sample and a few weeks later they give you access to a private web page that will show you a breakdown of the types of bacteria they found.</p>
<div id="how-to-use-the-ubiome-taxonomy-files" class="section level2">
<h2><span class="header-section-number">3.1</span> How to Use the uBiome Taxonomy Files</h2>
<p>The information you see in the uBiome web app is a greatly simplified view of the data torrent that comes from the gene sequencing machines that crunched your sample. Your raw data can be thousands of lines long, so to be more useful, uBiome provides a much more concise version. It’s called a “taxonomy” file, which is generally only a few hundred lines showing just the organisms that the uBiome computer algorithms think they found in your sample.</p>
<p>Because uBiome JSON files are well-structured, it’s easy to process them with other software, including Excel. That’s our next step: bring this data into Excel.</p>
<div id="study-your-results-in-excel" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Study your results in Excel</h3>
<p>The uBiome web interface includes an easy way to download your data as a CSV (comma-separated value) file, which can be read into a spreadsheet like Excel. Just click the ‘downloads’ in the ‘Advanced’ tab, like this:</p>
<p><img src="assets/ubiomeExplorerAdvancedTab.jpg" /><!-- --></p>
<p>Then select “Download Taxonomy CSV” <img src="assets/ubiomeExplorerAdvancedTab.jpg" /><!-- --></p>
<p>and a CSV copy of your results will be downloaded to your computer.</p>
<p>Once the data is in Excel, you can work with it just as you would with any Excel data sheet. Here’s how one of my recent uBiome taxonomy files looks in Mac Excel:</p>
<p><img src="HowToAnalyze/ExcelScreenOverview.png" width="919" /></p>
<p>At this point, we can now apply the full power of your Excel skills to do the analysis. I usually start by turning the sheet into a “table”, to make it easier to do filtering and sorting later. Most popular spreadsheets have an easy way to do this. On Mac Excel, I do that from the “insert” menu:</p>
<p><img src="HowToAnalyze/images/media/ubiomeExcelInsertTable.jpg" /><!-- --></p>
<p>Once I have a table, then it’s easy to apply a filter. For now I just want to look at the Phylum level, so I click the tiny triangle to the right of the “tax_rank” label, like this:</p>
<p><img src="HowToAnalyze/images/media/ubiomeExcelFilterTable.jpg" /><!-- --></p>
<p>And now I can look just at the Phylum rank. I won’t bother showing each step after this, but after filtering and sorting, I can quickly end up with something like this:</p>
<p><img src="HowToAnalyze/images/media/ubiomeExcelSorted.jpg" /><!-- --></p>
<p>Looks like there is about 80% Firmicutes in this sample, with Bacteroidetes at about 12%, and a few other phyla under about 4%. We’ll go into more detail about what this means, but first let’s talk a little about the uBiome taxonomy file format.</p>
</div>
<div id="ubiome-json-taxonomy-files" class="section level3">
<h3><span class="header-section-number">3.1.2</span> uBiome JSON taxonomy files</h3>
<p>Excel is great for a quick overview of your sample, but to really take advantage of the raw data it helps to understand the raw JSON format.</p>
<p>After uBiome’s lab completes the sequencing of your sample, the data generated goes into a bioinformatics pipeline where powerful algorithms use the zillions of DNA letters to find names for the specific organisms that inhabit you. One result, after many hours of computation, is a list of the organisms (taxa) found, along with their taxonomical rank and abundance. The result is stored in a simple structured text format called JSON (JavaScript Object Notation), commonly used across the web. Although it’s not as convenient for people as the rest of the uBiome web site, programmers refer to JSON as “human readable”, because if you squint enough you can sorta tell what it means without a computer program. Here’s a sample:</p>
<pre><code>{
  &quot;ubiome_bacteriacounts&quot;: [{
    &quot;taxon&quot;: &quot;2&quot;,
    &quot;parent&quot;: &quot;131567&quot;,
    &quot;count&quot;: &quot;22691&quot;,
    &quot;count_norm&quot;: &quot;1000000&quot;,
    &quot;avg&quot;: null,
    &quot;tax_name&quot;: &quot;Bacteria&quot;,
    &quot;tax_rank&quot;: &quot;superkingdom&quot;,
    &quot;tax_color&quot;: null
  }, {
    &quot;taxon&quot;: &quot;1239&quot;,
    &quot;parent&quot;: &quot;2&quot;,
    &quot;count&quot;: &quot;15414&quot;,
    &quot;count_norm&quot;: &quot;679300&quot;,
    &quot;avg&quot;: null,
    &quot;tax_name&quot;: &quot;Firmicutes&quot;,
    &quot;tax_rank&quot;: &quot;phylum&quot;,
    &quot;tax_color&quot;: &quot;5E6591&quot;
  }</code></pre>
<p>JSON is just structured data. That’s it: a big bunch of lookup keys and values. The keys are unique identifiers and the values are their values. The whole file looks that way. The only requirement of a JSON file is that it be precisely consistent with this pattern, because after all it will need to be read by a computer eventually.</p>
<p>The uBiome Taxonomy JSON files include the key <code>uBiome_bacteriacounts</code> which is mapped to a bunch of fields, each representing a particular organism found in your sample. Look at the figure above to get the basic idea.</p>
</div>
<div id="the-ubiome-data-tags" class="section level3">
<h3><span class="header-section-number">3.1.3</span> The uBiome data tags</h3>
<p>Let’s review the meaning of each of the columns, which are derived from the following uBiome JSON keys:</p>
<p><strong>count</strong>: an absolute measure of number of organisms found in the sample. Without knowing the size of the sample, or how many times the DNA inside was processed through PCR amplification, this number doesn’t mean much except in relation to other counts at the same taxonomical rank.</p>
<p><strong>count_norm</strong>: a “normalized” version of the count, created by simply dividing the count on a given row by the count found on the row with tax_name = Bacteria and then multiplying by one million. It’s easier to just think of it as parts per million: each unit is 1 / 10,000th of a percent. For example, if you see a row with count_norm = 500,000, you can just think of that as 50% of the sample.</p>
<p><strong>tax_name</strong>: this is the classification of the organism based on the level of its taxonomy. If you were looking at a human being, for example, you would see homo sapiens if you selected tax_rank = species, but you’d see mammalia if you selected tax_rank = class.</p>
<p><strong>tax_rank</strong>: tells the level of the taxonomy. In daily conversation about animals or plants, we usually refer to the species (e.g. homo sapiens), but sometimes it’s more useful to talk about bigger groupings of related organisms. For example, humans are members of the class mammalia, along with tigers and horses. If this spreadsheet were counting organisms at the level of class mammalia, the count_norm would almost certainly be bigger than the count_norm for humans alone, unless humans were the only type of mammal found in the sample.</p>
<p><strong>taxon</strong> and <strong>parent</strong>: these help identify the ranking in a more precise way by pointing out which tax_ranks are subsets of which. For example, Bacteroidia above has a parent = 976, meaning that it is a subset of the taxon 976, Bacteroidetes. When you follow the various taxons and parents up the chain, you’ll see they all end in the superkingdom Bacteria, which has a taxon of 2. The values for these numbers, incidentally, are taxonomical numbers from the curated database at NCBI, the national bioinformatics center run by the U.S. government. Enter the number into the taxonomy browser at the NCBI Taxonomy Browser and you can learn as much as you want about that organism.</p>
<p><strong>tax_color</strong> doesn’t matter for this anaysis, but uBiome software uses this to colorize their pretty graphs to make them more readable.</p>
<p>Of these fields, the most important ones are <strong>tax_name</strong>, <strong>tax_rank</strong>, and <strong>count_norm</strong>.</p>
</div>
<div id="converting-ubiome-taxonomy-json-to-excel" class="section level3">
<h3><span class="header-section-number">3.1.4</span> Converting uBiome taxonomy JSON to Excel</h3>
<p>As already noted, it’s easy to bring uBiome taxonomy data into Excel, and that’s fine for just one or two samples. But if you want more powerful analysis, you’ll need the raw JSON. Go to the <a href="http://explorer.ubiome.com">uBiome Explorer page</a> and click on the section labeled “Advanced” and then “Downloads”:</p>
<p><img src="assets/ubiomeExplorerDownloads.jpg" /><!-- --></p>
<p>You’ll see a page of JSON representing what uBiome found in your sample.</p>
<p><img src="HowToAnalyze/uBiomeScreenShotTaxonomyJSON.png" width="471" /></p>
<p>On a desktop computer, if you select-all (press control-A or command-A) then you can copy this data to the clipboard and then into an email to share with a friend.</p>
<p>If you like, you can even convert the JSON directly into Excel. Google “Json to Excel” or “Json to CSV” and you’ll find one. Just paste your data into one of those sites to convert it for either Excel’s native XLS or XLSX format, or the more universal CSV format, readable by Excel. Here are a few of the sites I’ve tested:</p>
<p><a href="http://www.convertcsv.com/json-to-csv.htm" class="uri">http://www.convertcsv.com/json-to-csv.htm</a></p>
<p><a href="http://www.json-xls.com/json2xls" class="uri">http://www.json-xls.com/json2xls</a></p>
<p>It’s nice to know that we can always go back to a spreadsheet version when needed. But the real power of JSON comes when you want to handle more than just one or two samples.</p>
<hr />
</div>
</div>
<div id="comparing-samples" class="section level2">
<h2><span class="header-section-number">3.2</span> Comparing samples</h2>
<p>The uBiome JSON taxonomy file is extremely helpful at giving you a short summary of what you care about most in a sample: the organisms within it and their relative amounts. But even with these annotations, there’s a limit to what you can learn in a single sample. To be able to start hacking your microbiome, you will need to learn how to compare multiple samples.</p>
<p>There are two main things you’ll want to measure when comparing two samples:</p>
<p><strong>Uniqueness</strong>: which organisms are found in only one sample and not the other? If you are comparing two samples from the same individual (e.g. Yourself), then uniqueness is another way of talking about either extinctions (when a taxon has disappeared entirely in a later sample) and appearances (when a new taxon magically shows up). In other words, you care a lot about the order in which the samples were taken.</p>
<p>On the other hand, if you’re comparing two separate individuals, then the ordering of the samples doesn’t matter. Uniqueness just tells which organisms are unique to a specific person.</p>
<p><strong>Relative abundance</strong>: which person has more of which organism? There are two senses in which we care about the relative frequency of an organism’s occurrence in a sample. You might care, among all the taxons found, which ones are more highly represented in absolute amounts? Or you might care on a relative basis. To help understand why this matters, it may help to think of the following specific example:</p>
<p>In many Americans the phylum Firmicutes makes up a majority of the sample, often reaching 60% or more. You might find two people, one who has 60% Firmicutes and the other with 30% — only half as much. The relative difference is 2x and the absolute difference is 30 percentage points.</p>
<p>On the other hand, you might find Bifidobacterium makes up 10% of the first sample and 5% of the second. In this case, the relative difference is 2x — just like the Firmicutes case — but the absolute difference is only 5 percentage points, much less than in the first example.</p>
<p>The built-in uBiome web tools look for relative differences. A taxon might be a minuscule component of both samples, but when you start from a small base, even a few additional organisms can make up a big percentage change. In the extreme case, a sample with a count_norm of 100 compared to another sample with a count_norm of 300 will show a 3X increase — very high in aBiome terms, even though the absolute difference is only 300 - 100 = 200 organisms.</p>
<p>Compare that with a an organism with a count_norm of 50,000 in one sample and 51,000 in another sample. UBiome’s algorithms will treat these as a virtual tie, even though the absolute number of organisms in the second sample outnumber the first by 1,000 organisms.</p>
<p>Which method of measuring is better? It depends on what you care more about. Some species need only a tiny representation in a sample to make a big difference in health outcomes.</p>
<p>I prefer to use both methods of measurement: absolute number changes for the most popular taxons, and relative amounts more for the less-popular ones.</p>
<hr />
</div>
<div id="make-a-single-sheet-with-all-your-samples" class="section level2">
<h2><span class="header-section-number">3.3</span> Make a single sheet with all your samples</h2>
<p>I often read news about a fresh scientific discovery involving the microbiome and immediately wonder if the discovery applies to me. For example, I recently saw a study from Oregon State University]^[<a href="http://www.sciencedaily.com/releases/2015/06/150622182034.htm" class="uri">http://www.sciencedaily.com/releases/2015/06/150622182034.htm</a> that seemed to find a link between high sugar diets and “cognitive flexibility”, i.e. your ability to adapt and adjust to changing circumstances. The study’s author, Kathy Magnusson, a professor in the OSU College of Veterinary Medicine, found that mice who eat lots of sugar have elevated levels of Clostridiales bacteria, and that this seemed to relate to a slower ability to solve a maze. Hmmm, I thought — how much Clostridiales do I have?</p>
<p>If you have just one uBiome result, that’s easy: log into <a href="http://explorer.ubiome.com" class="uri">http://explorer.ubiome.com</a> and search for it in the section “All My Bacteria”. But in my experience a single result doesn’t tell you much. You really need at least two and hopefully several uBiome results to see what might be actionable. In my case, I want to know how my Clostridiales may have changed over time.</p>
<p>If you are comfortable with programming, uBiome has an opensource Python script to generate a single Excel table that combines multiple separate JSON files. Something like <a href="https://github.com/ubiome-opensource/microbiome-data/blob/master/sprague/spragueResultsthruJun2015.xlsx">this</a>:</p>
<p><img src="HowToAnalyze/images/media/allResultsOneTable.jpeg" /><!-- --></p>
<p>The data makes it easy to generate a chart showing how my Clostridiales changes over time:</p>
<p><img src="HowToAnalyze/images/media/spragueClostridiales.jpeg" /><!-- --></p>
<p>Hmmm, in my case it looks like something happened since last fall to increase my Clostridiales levels. Maybe it was the potato starch I tried in order to <a href="http://www.ubiomeblog.com/hacking-sleep/">hack my sleep</a>? Was it my <a href="http://www.ubiomeblog.com/my-microbiome-in-the-jungle/">trip to Central America</a> in February? And of course the biggest question: has the increase affected my cognitive flexibility? I’m <a href="http://blog.richardsprague.com/2015/07/my-qs15-slides.html">not really sure</a>. Whatever happened, the level of Clostridiales seems to have stabilized in the past couple of months.</p>
<p>uBiome has identified more than 900 unique taxa (groups of organisms) in the half-dozen samples I’ve submitted over the past year, and after running this script I have them all laid out on a single page. Now, armed with this one spreadsheet I can search anytime for a new microbe and quickly see if I have it now, or if it’s ever been detected in a previous test. Reading news about microbiome has taken on a whole new personal meaning when I can see if the discovery relates to me.</p>
<p>Next we’ll show you how to make your own charts.</p>
</div>
<div id="analyze-your-ubiome-results-in-python" class="section level2">
<h2><span class="header-section-number">3.4</span> Analyze your uBiome Results in Python</h2>
<p>The charts and tables you see in the <a href="https://app.ubiome.com/explore">uBiome Explorer</a> are generated from the same data file that you can download for yourself in the Advanced tab. If you’re comfortable with programming, you can pull that data into your favorite software tools for even more analysis.</p>
<p>To get started, look at the open source <a href="https://github.com/ubiome-opensource/microbiome-tools">uBiome Tools Repository</a> which contains source code that will let you:</p>
<ul>
<li><p>Compare two samples to find the unique organisms in each.</p></li>
<li><p>Compare two samples and show the differences in counts for each.</p></li>
<li><p>Merge more than one sample into a single file.</p></li>
</ul>
<p>For Python programmers, the code is available either in the official uBiome library or as an easy-to-install library hosted on the <a href="https://pypi.python.org/pypi/ubiome">Python Package Index</a> and installable through PIP. Just type this from the console application on your computer (either Terminal on Mac, or Powershell on Windows).</p>
<pre><code>$ pip install ubiome</code></pre>
<p>The uBiome package is designed for either Python 2 or 3, and once installed you can use it as a console script, for a new system command like this:</p>
<pre><code>&gt; ubiome -h

usage: ubiome.py [-h] [-c COMPARE] [-u UNIQUE] [-d DEBUG] sample1 sample2

positional arguments:
  sample1               filename for a valid uBiome JSON taxonomy file
  sample2               sample you are comparing to

optional arguments:
  -h, --help            show this help message and exit
  -c COMPARE, --compare COMPARE
                        Compare sample1 with with sample2
  -u UNIQUE, --unique UNIQUE
                        Find items in sample1 not in sample2
  -d DEBUG, --debug DEBUG
                        turn debug mode to run tests</code></pre>
<p>If you have two uBiome-formatted JSON files in the current directory, try this:</p>
<pre><code>$ ubiome -c sample1.json sample2.json</code></pre>
<p>It will compare the two samples and output a comma-separated (CSV) dump of the differences between them. Similarly, the following command returns the parts of a sample that are unique:</p>
<pre><code>$ ubiome -u sample1.json sample2.json &gt; uniques.csv</code></pre>
<p>Now you’ll have a new file in your directory, “uniques.csv”, that you can load into any spreadsheet program. The file will have a series of rows representing each taxa of bacteria that are unique, i.e. only found in “sample1.json”.</p>
<p>You’ll find both the compare and unique commands useful for understanding the differences between two of your own samples. If you have a sample of yourself before taking probiotics, and another afterwards, for example, this can help you tell what organisms changed and by how much.</p>
<p>The uBiome package can of course be imported into Python as the basis of your own scripts and modules. To see how, run the following series of commands in your Python 2+ or 3+ console:</p>
<pre><code>$ python

&gt;&gt;&gt; from ubiome import *
&gt;&gt;&gt; x1 = UbiomeSample(&quot;sample1.json&quot;)
&gt;&gt;&gt; x2 = UbiomeSample(&quot;sample2.json&quot;)
&gt;&gt;&gt; x = UbiomeMultiSample(x1)
&gt;&gt;&gt; x.merge(x2)
&gt;&gt;&gt; # continue to merge as many additional samples as you like
&gt;&gt;&gt; x.write(&quot;x.csv&quot;)</code></pre>
<p>Now your directory will have a new file ‘x.csv’ with all samples merged. The first row is all the taxons ever found in your samples, and the other columns are your different samples, with rows containing the <strong>count_norm </strong>value for every taxon, including zero values for those taxa that are missing in a particular test.</p>
<p>The merged form, called a <strong>UbiomeMultiSample</strong>, can do even more. For example, continue to merge a few more samples in the example above, and then run this code:</p>
<pre><code>samples = x.originalSampleObjects # just the UbiomeSample objects of the merged items

diversity = [sample.diversity() for sample in samples]

dates = [sample.date for sample in samples]</code></pre>
<p>If you have matplotlib on your system and run this:</p>
<pre><code>plot(dates,diversity)</code></pre>
<p>You will see this chart that plots the diversity of a series of samples over time:</p>
<p><img src="HowToAnalyze/images/media/uBiomeDiversityChart.jpg" /><!-- --></p>
<p>See the script <a href="https://github.com/ubiome-opensource/microbiome-tools/blob/master/ubiome_example.py">ubiome_example.py</a> for a complete example. There are also several user-supplied JSON files for you to try if you want to test your own code on real samples.</p>
</div>
<div id="analyze-a-fastq-file" class="section level2">
<h2><span class="header-section-number">3.5</span> Analyze a FASTQ file</h2>
<p>The uBiome JSON taxonomy information will be more than enough for most people new to biology. But for BioCoders and anyone interested in expanding their bioinformatics skills, the raw data files are the place to look.</p>
<p>FASTQ files are the text-based results of a run at your sample using one of Illumina’s high-end gene sequencers.</p>
<p>For example, the first line in my data is:</p>
<pre><code>@NS500457:29:H2HLGAFXX:1:11101:16760:1033 1:N:0:ACTTACA+NTGCGAA</code></pre>
<p>This contains information about the gene sequencing hardware uBiome used for this sample. The other lines in the raw text file contain the actual base pairs detected by the sequencer, along with quality information to indicate the certainty of a particular read. In my case, the entire file is over 170,000 lines long, and that’s just one of the eight files in the FASTQ archive. That’s a lot of data!</p>
<p>Although it’s possible to analyze this data on my own in a tool like BioPython, it’s much easier to submit the raw data to a public gene processing server, such as the Metagenomics Analysis Server (MG RAST) hosted by Argonne National Labs at this site:</p>
<p><a href="http://metagenomics.anl.gov/" class="uri">http://metagenomics.anl.gov/</a></p>
<p>Note: You must use the Firefox browser (not Chrome, IE, or Safari), and before submitting any jobs you must apply for a (free) login ID, which you receive by email in a day or two.</p>
<p>After logging in with your new account ID, click the “upload” icon on the home page and follow the process to prepare and submit your data:</p>
<p><img src="HowToAnalyze/images/media/image1.jpeg" /><!-- --></p>
<ol style="list-style-type: decimal">
<li><p>Ignore the part about preparing the meta data; you can add it later.</p></li>
<li><p>Upload files. The site offers to let you upload from the web browser, but I’ve had better luck using their api and the following curl command:</p></li>
</ol>
<pre><code>curl -H &quot;auth: webkey&quot; -X POST -F &quot;upload=@/path\_to\_file/metagenome.fasta&quot; &quot;&lt;http://api.metagenomics.anl.gov/1/inbox/&gt;&quot; &amp;gt; curl\_output.txt</code></pre>
<p>I just type this from Terminal on my Mac. Note that you’ll need to get your own webkey (hit the ‘view webkey’ button on the submission page) which you will substitute into the curl command line above. Also change the filename part of the “upload=@” section to point to the correct file. Note that uBiome fastq files are compressed, but it’s okay to upload them as-is. You’ll uncompress them on the server after the upload.</p>
<ol style="list-style-type: decimal">
<li>Manage your inbox. Once the file has successfully uploaded, you’ll see it in your inbox. At this point, you should “unpack selected” if you uploaded a compressed file. Note depending on how busy the server is, it may be minutes or even hours of waiting, and pressing “update inbox” before you see the File Information shown below.</li>
</ol>
<p><img src="HowToAnalyze/images/media/image2.png" width="660" /></p>
<p>Next you’re ready to submit the data</p>
<p><img src="HowToAnalyze/images/media/image3.jpeg" /><!-- --></p>
<p>Here are my suggestions for this section</p>
<ol style="list-style-type: decimal">
<li><p>Check the box to say you won’t supply metadata.</p></li>
<li><p>Create a new project and select it.</p></li>
<li><p>Select the files you uploaded</p></li>
<li><p>Use the default values to choose pipeline options</p></li>
<li><p>Submit the job, being sure to make the data publicly accessible immediately, to increase the priority.</p></li>
</ol>
<p>There you go! Now a job has been submitted and you wait for it to complete. It usually takes a day or two to complete, after which you’ll receive an email notification. After that, click the bar chart icon on the upper right of the page:</p>
<p><img src="HowToAnalyze/images/media/image4.jpeg" /><!-- --></p>
<p>This brings you to an analysis page where you can study your data in much more detail.</p>
<p>When I select my data under “Metagenomes” on the analysis page, I’m offered several options for Annotation Sources. Since the uBiome fastq data comes from 16S ribosomal RNA, I choose to compare my data against one of the large databases of known RNA. The popular Greengenes database is one of them, so I select it, and then ask to generate a table of the results.</p>
<p><img src="HowToAnalyze/images/media/image5.png" width="435" /></p>
<p>The table, sorted by abundance and filtered to phylum, looks like this:</p>
<p><img src="HowToAnalyze/images/media/image6.png" width="311" /></p>
<p>The results roughly conform to what we see from uBiome: the most common organisms are the same in each case, and although the scales are different, the abundance computed corresponds roughly to what we see from the uBiome count field.</p>
<p>Other MG RAST tools let you compare your data against other databases of bacteria and proteins, graph the results, and perform sophisticated statistical analysis to determine other unique features of your sample. You can also compare your sample to microbiomes that have been uploaded by others, enabling you to study your results in much more detail.</p>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="17">
<li id="fn17"><p><a href="https://www.indiegogo.com/projects/ubiome-sequencing-your-microbiome#/" class="uri">https://www.indiegogo.com/projects/ubiome-sequencing-your-microbiome#/</a><a href="explore-your-microbiome.html#fnref17">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="science.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="my-experiments.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/richardsprague/mhg/edit/master/exploreYourMicrobiome.Rmd",
"text": "Edit"
},
"download": ["Microbiome Hackers Guide (Bookdown Edition).pdf", "Microbiome Hackers Guide (Bookdown Edition).epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
